text/XXX-template-poly.md

```diff
@@ -0,0 +1,154 @@
+- Title: Template poly redesign using sort poly
+
+- Drivers: Gaëtan Gilbert
+
+----
+
+# Summary
+
+Template poly currently provides 2 features in 1: implicit universe
+instances and adhoc sort polymorphism handling only Prop/Type.
+
+We want to make it less adhoc and base it on more well understood
+systems such as regular sort poly.
+
+This should also allow to stop producing useless constraints (like u <= prod.u0)
+when a template inductive is fully applied.
+
+# Detailed design
+
+## Declaration
+
+An inductive may be declared to use implicit univ instances when:
+
+- it is non mutual (nested is ok) (this rule is in master but maybe we will remove it)
+- its univ declaration has only univ variables unbounded from below
+- its univ declaration has at most 1 qvar variable.
+  If it has one, it is the quality of the output sort.
+- each parameter type either does not mention the bound univs (CLOSED),
+  or is of the form `forall args, Type@{q|u}` (BINDING)
+  where the `args` types do not mention the bound univs and qvar,
+  `q` is either Type or the unique bound qvar,
+  and `u` is either constant or one of the bound univ levels.
+- the indices types and constructor types do not mention the bound univs and qvar.
+
+NB: the qvar (if there is one) must be "above prop" so should not appear in relevance marks
+(it's always `Relevant`) so eg
+~~~
+Inductive prod@{q|u v|} (A:Type@{q|u}) (B:Type@{q|v}) : Type@{q|max(u,v)} :=
+  pair : A -> B -> prod A B.
+~~~
+is accepted and the arrow `A -> ...` has relevance mark `Relevant` not `RelevanceVar q`.
+See also "Note on nested templates and above Prop" below.
+
+NB': these rules mean the bound univs have `Irrelevant` cumulativity
+variance (not to be confused with relevance).
+
+## Instantiation
+
+Given a default instance (with default quality `Type`),
+a (possibly partial) application of the inductive or its constructors
+to parameters is handled by:
+
+For each BINDING parameter, we obtain the universe level (which may be algebraic)
+and quality from the actual passed parameter if available, otherwise from the default instance.
+The quality (if bound) must be "above Prop".
+
+If the quality is bound (ie non constant), it is assigned to the max of the inferred qualities
+(this is possible because of the "above Prop" assumption).
+For each bound univ level, it is assigned to the max of the corresponding obtained levels.
+
+This gives the implicit instance.
+We then check any constraints on the variables and do regular typechecking.
+
+## Subject reduction
```

Comment:
TODO check that no problem appears with constructors

Reply:
also matches

Reply:
A priori the same must be done for (partial applications of) constructors otherwise typechecking their parameter arguments them will naturaly add constraints on the default instance.